oximeter/db/src/oxql/table.rs

```diff
@@ -0,0 +1,293 @@
+    }
+
+    /// Get a timeseries matching the provided key, if any.
+    pub fn get_mut(&mut self, key: TimeseriesKey) -> Option<&mut Timeseries> {
+        self.timeseries.get_mut(&key)
+    }
+
+    /// Insert a new timeseries into the table.
+    ///
+    /// If the timeseries already exists, an error is returned. Use
+    /// [`Table::replace()`] to replace an existing timeseries.
+    ///
+    /// It is an error of the timeseries does not have the same schema as the
+    /// others in the table (if any).
+    pub fn insert(&mut self, timeseries: Timeseries) -> Result<(), Error> {
+        anyhow::ensure!(
+            self.matches_schema(&timeseries),
+            "Timeseries in a table must have the same schema",
+        );
+        let key = timeseries.key();
+        let Entry::Vacant(e) = self.timeseries.entry(key) else {
+            return Err(anyhow::anyhow!(
+                "Timeseries with key {} already exists",
+                key,
+            ));
+        };
+        e.insert(timeseries);
+        Ok(())
+    }
+
+    /// Replace a timeseries in the table.
+    pub fn replace(&mut self, timeseries: Timeseries) {
+        let key = timeseries.key();
+        let _ = self.timeseries.insert(key, timeseries);
+    }
+
+    /// Add multiple timeseries to the table.
+    ///
+    /// An error is returned if any timeseries already exist.
+    pub fn extend(
+        &mut self,
+        timeseries: impl Iterator<Item = Timeseries>,
+    ) -> Result<(), Error> {
+        for t in timeseries {
+            self.insert(t)?;
+        }
+        Ok(())
+    }
+
+    /// Return the number of timeseries in the table.
+    pub fn len(&self) -> usize {
+        self.timeseries.len()
+    }
+
+    /// Return a mutable iterator over timeseries in the table.
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut Timeseries> {
+        self.timeseries.values_mut()
+    }
+
+    /// Return an iterator over timeseries in the table.
+    pub fn iter(&self) -> impl Iterator<Item = &Timeseries> {
+        self.timeseries.values()
+    }
+
+    /// Consume the table and return an iterator over its timeseries.
+    pub fn into_iter(self) -> impl Iterator<Item = Timeseries> {
+        self.timeseries.into_values()
+    }
+
+    /// Return `true` if all the timeseries in this table are aligned, with the
+    /// same alignment information.
+    ///
+    /// If there are no timeseries, `false` is returned.
+    pub fn is_aligned(&self) -> bool {
+        let mut timeseries = self.timeseries.values();
+        let Some(t) = timeseries.next() else {
+            return false;
+        };
+        let Some(alignment) = t.alignment else {
+            return false;
+        };
+        timeseries.all(|t| t.alignment == Some(alignment))
+    }
+
+    /// Return the alignment of this table, if all timeseries are aligned with
+    /// the same alignment.
+    pub fn alignment(&self) -> Option<Alignment> {
+        if self.is_aligned() {
+            Some(
```

Comment:
Perhaps drop the `Some` wrapper and the final unwrap?

Reply:
I kept that intentionally, to catch a programming error where `is_aligned()` returns `true`, but the timeseries itself doesn't have an alignment for some reason.